src/three-components/Shadow.ts

```diff
@@ -13,26 +13,103 @@
  * limitations under the License.
  */
 
-import {Mesh, PlaneGeometry, ShadowMaterial} from 'three';
+import {DirectionalLight, Mesh, PlaneBufferGeometry, ShadowMaterial} from 'three';
 
-/**
- * Creates a mesh that can receive and render shadows.
- */
-export default class Shadow extends Mesh {
-  /**
-   * Create a shadow mesh.
-   */
-  constructor() {
-    const geometry = new PlaneGeometry(2000, 2000);
-    geometry.rotateX(-Math.PI / 2);
-
-    const material = new ShadowMaterial(({
-      color: 0x111111,
-      opacity: 0.2,
-    } as any));
-
-    super(geometry, material);
-    this.receiveShadow = true;
-    this.name = 'ShadowMesh';
+import ModelScene from './ModelScene';
+
+// Nothing within shadowOffset of the bottom of the model casts a shadow
+// (this is to avoid having a baked-in shadow plane cast its own shadow).
+const SHADOW_OFFSET = 0.001;
+const BASE_SHADOW_OPACITY = 0.1;
+// The softness of the shadow is controlled with this resolution parameter. The
+// lower the resolution, the softer the shadow.
+const MAX_SHADOW_RESOLUTION = 64;
+
+export class Shadow extends DirectionalLight {
+  private shadowMaterial = new ShadowMaterial;
+  private floor: Mesh;
+  public needsUpdate = false;
+
+  constructor(private scene: ModelScene) {
+    super();
+
+    // We use the light only to cast a shadow, not to light the scene.
+    this.intensity = 0;
+    this.castShadow = true;
+
+    this.floor = new Mesh(new PlaneBufferGeometry, this.shadowMaterial);
+    this.floor.receiveShadow = true;
+    this.floor.castShadow = false;
+    scene.pivot.add(this.floor);
+    scene.pivot.add(this);
```

Comment:
This is an inverted relationship between a scene and one of its descendent nodes. Let's change it so that the `ModelScene` creates a new `Shadow` and appends it to the `pivot`.